estate/models/estate_property.py

```diff
@@ -0,0 +1,131 @@
+            ("offer received", "Offer Received"),
+            ("offer accepted", "Offer Accepted"),
+            ("sold", "Sold"),
+            ("canceled", "Canceled"),
+        ],
+        tracking=True,
+    )
+    salesman_id = fields.Many2one(
+        "res.users", string="Salesman", default=lambda self: self.env.user
+    )
+    buyer_id = fields.Many2one("res.partner", string="Buyer", copy=False)
+    property_type_id = fields.Many2one("estate.property.type", string="Property Type")
+    tag_ids = fields.Many2many("estate.property.tag")
+    offer_ids = fields.One2many("estate.property.offer", "property_id")
+    total_area = fields.Float(compute="_compute_total_area")
+    best_price = fields.Float(compute="_compute_best_offer")
+
+    _sql_constraints = [
+        (
+            "positive_expected_price",
+            "CHECK(expected_price > 0.0)",
+            "A property expected price must be strictly positive",
+        ),
+        (
+            "positive_selling_price",
+            "CHECK(selling_price >= 0.0)",
+            "A property selling price must be positive",
+        ),
+    ]
+
+    @api.constrains("selling_price", "expected_price")
+    def _check_selling_price(self):
+        for record in self:
+            if not float_is_zero(record.selling_price, precision_digits=2):
+                if (
+                    float_compare(
+                        record.selling_price,
+                        0.9 * record.expected_price,
+                        precision_digits=2,
+                    )
+                    == -1
+                ):
+                    raise ValidationError(
+                        "Selling price cannot be lower than 90% of the expected price"
+                    )
+
+    @api.depends("living_area", "garden_area")
+    def _compute_total_area(self):
+        for record in self:
+            record.total_area = record.living_area + record.garden_area
+
+    @api.depends("offer_ids.price")
+    def _compute_best_offer(self):
+        for record in self:
+            if record.offer_ids:
+                record.best_price = max(record.offer_ids.mapped("price"))
+            else:
+                record.best_price = 0
+
+    @api.onchange("garden")
+    def _onchange_garden(self):
+        if self.garden:
+            self.garden_area = 10
+            self.garden_orientation = "north"
+        else:
+            self.garden_area = 0
+            self.garden_orientation = ""
+
+    def action_cancel_property(self):
+        for record in self:
+            if record.state == "sold":
+                raise UserError(_("Sold property can not be canceled."))
+            else:
+                self.write({"state": "canceled"})
+        return True
+
+    def action_sold_property(self):
+        for record in self:
+            if record.state == "canceled":
+                raise UserError(_("Canceled property can not be sold."))
+            else:
+                self.write({"state": "sold"})
```

Comment:
```suggestion
                record.state = "sold"
```
Try to avoid direct usage of write method when we are updating a single field. check in other files also
And we have to use record here instead of self as we are looping through self using record.